### DISCUSSION TITLE
PP-OCRv4_server_rec_doc 怎么转onnx, 怎么用之前的单独字典文件推理？

### DISCUSSION THREAD

**Discussion by nissansz** [PaddlePaddle/PaddleOCR#14973] [Q&A]
PP-OCRv4_server_rec_doc 怎么转onnx, 怎么用之前的单独字典文件推理？

PP-OCRv4_server_rec_doc https://paddlepaddle.github.io/PaddleX/latest/module_usage/tutorials/ocr_modules/text_recognition.html#_2

**Comment by GreatV** [comment]
要将 PP-OCRv4_server_rec_doc 模型转换为 ONNX，并使用你之前的独立字典文件进行推理，可以按照以下步骤操作：

一、模型转换为 ONNX

PP-OCR 使用 Paddle 框架，因此转换为 ONNX 需要借助 Paddle2ONNX 工具。在 PaddleOCR 的部署文档中提到可以通过 ./paddle2onnx 目录下的方法完成转换。具体步骤如下：

1. 安装 paddle2onnx

你可以使用如下命令安装：

pip install paddle2onnx

2. 准备 inference 模型

确保你已经导出了 PP-OCRv4_server_rec_doc 的 Paddle inference 模型（含 model.pdmodel 和 model.pdiparams）。通常可以通过 tools/export_model.py 脚本进行导出：

python3 tools/export_model.py \
    -c configs/rec/PP-OCRv4/rec_svtrnet_server_mv3.yml \
    -o Global.pretrained_model=路径/PP-OCRv4_server_rec_doc/best_accuracy \
    Global.save_inference_dir=output/rec_infer

3. 使用 paddle2onnx 进行转换

转换命令如下：

paddle2onnx \
  --model_dir output/rec_infer \
  --model_filename inference.pdmodel \
  --params_filename inference.pdiparams \
  --save_file ppocrv4_rec.onnx \
  --opset_version 11 \
  --enable_onnx_checker True

你可以根据实际文件名调整路径。

二、使用独立字典文件推理

在 PaddleOCR 中，字典文件（通常为 txt 格式）中每一行代表一个字符。ONNX 推理时你需要将其重新加载到推理脚本中以还原识别结果。

1. 加载字典

在推理代码中，加载你的独立字符字典文件，例如：

def load_dict(dict_path):
    chars = []
    with open(dict_path, 'r', encoding='utf-8') as f:
        for line in f:
            chars.append(line.strip())
    return chars

2. ONNX 模型推理后输出通常为 logits（概率分布），你需要结合字典将其转换为字符串。一般会结合 CTC 解码或者其他解码方式将预测索引映射为字符。

举个例子：

import numpy as np

def decode(output, dict_chars):
    result = ""
    last_index = -1
    for i in np.argmax(output, axis=1):
        if i != last_index and i != len(dict_chars):  # len(dict_chars) 为 CTC blank
            result += dict_chars[i]
        last_index = i
    return result

注意：ONNX 推理通常使用 ONNXRuntime 等工具进行，输入必须按照模型预处理处理图像。

三、推荐文档链接

你可以参考 PaddleOCR 官方提供的 ONNX 导出与部署文档：

👉 Paddle2ONNX 推理：https://github.com/PaddlePaddle/PaddleOCR/blob/dygraph/deploy/paddle2onnx/readme_ch.md

如果你使用的是 tools/export_model.py 创建 inference 模型，建议参考：

👉 导出模型文档：https://github.com/PaddlePaddle/PaddleOCR/blob/dygraph/doc/doc_ch/inference_ppocr.md#2ppocr%E6%A8%A1%E5%9E%8B%E5%AF%BC%E5%87%BA

总结：

- 使用 paddle2onnx 将 Paddle 模型转换为 ONNX；
- 加载你自己的字典文件，并在推理后将输出索引映射为字符；
- 编写完整 ONNX 推理代码时，注意对输入图像进行相同预处理，并正确设置解码逻辑。

如果你需要更详细的推理脚本，可提供你当前的使用框架（如 ONNXRuntime、TensorRT 等）。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

 ⚠️ This is an AI-generated response. Please verify any critical information.
